File: Arrays/Plus One/Python /recursive_solution.py
def plusOne(digits):
    """
    Recursive approach to handle carry.
    """

    def add_one(index):
        if index < 0:
            digits.insert(0, 1)
            return

        if digits[index] < 9:
            digits[index] += 1
        else:
            digits[index] = 0
            add_one(index - 1)

    add_one(len(digits) - 1)
    return digits


# Example usage
if __name__ == "__main__":
    print(plusOne([9]))       # [1, 0]
    print(plusOne([9, 9]))    # [1, 0, 0]

File: Arrays/Plus One/Cpp/recursive_solution.cpp
#include <bits/stdc++.h>
using namespace std;

void addOne(vector<int>& digits, int index) {
    if (index < 0) {
        digits.insert(digits.begin(), 1);
        return;
    }

    if (digits[index] < 9) {
        digits[index]++;
    } else {
        digits[index] = 0;
        addOne(digits, index - 1);
    }
}

vector<int> plusOne(vector<int>& digits) {
    addOne(digits, digits.size() - 1);
    return digits;
}

int main() {
    vector<int> digits = {9, 9};
    vector<int> result = plusOne(digits);

    for (int d : result)
        cout << d << " ";

    return 0;
}
